src/source-dpdk.c

```diff
@@ -0,0 +1,781 @@
+ */
+void TmModuleDecodeDPDKRegister(void)
+{
+    tmm_modules[TMM_DECODEDPDK].name = "DecodeDPDK";
+    tmm_modules[TMM_DECODEDPDK].ThreadInit = DecodeDPDKThreadInit;
+    tmm_modules[TMM_DECODEDPDK].Func = DecodeDPDK;
+    tmm_modules[TMM_DECODEDPDK].ThreadExitPrintStats = NULL;
+    tmm_modules[TMM_DECODEDPDK].ThreadDeinit = DecodeDPDKThreadDeinit;
+    tmm_modules[TMM_DECODEDPDK].cap_flags = 0;
+    tmm_modules[TMM_DECODEDPDK].flags = TM_FLAG_DECODE_TM;
+}
+
+static inline void DPDKDumpCounters(DPDKThreadVars *ptv)
+{
+    struct rte_eth_stats eth_stats;
+    int retval = rte_eth_stats_get(ptv->port_id, &eth_stats);
+    if (unlikely(retval != 0)) {
+        SCLogError(SC_ERR_STAT, "Failed to get stats for port id %d: %s", ptv->port_id,
+                strerror(-retval));
+        return;
+    }
+
+    uint64_t th_pkts = StatsGetLocalCounterValue(ptv->tv, ptv->capture_dpdk_packets);
+    StatsAddUI64(ptv->tv, ptv->capture_dpdk_packets, ptv->pkts - th_pkts);
+    SC_ATOMIC_ADD(ptv->livedev->pkts, ptv->pkts - th_pkts);
+
+    /* Some NICs (e.g. Intel) do not support queue statistics and the drops can be fetched only on
+     * the port level. Therefore setting it to the first worker to have at least continuous update
+     * on the dropped packets. */
+    if (ptv->queue_id == 0) {
+        StatsSetUI64(ptv->tv, ptv->capture_dpdk_drops, eth_stats.imissed + eth_stats.ierrors);
+        SC_ATOMIC_SET(ptv->livedev->drop, eth_stats.imissed + eth_stats.ierrors);
+    }
+}
+
+static void DPDKReleasePacket(Packet *p)
+{
+    int retval;
+    /* Need to be in copy mode and need to detect early release
+       where Ethernet header could not be set (and pseudo packet)
+       When enabling promiscuous mode on Intel cards, 2 ICMPv6 packets are generated.
+       These get into the infinite cycle between the NIC and the switch in some cases */
+    if (!PKT_IS_PSEUDOPKT(p) &&
+            (p->dpdk_v.copy_mode == DPDK_COPY_MODE_TAP ||
+                    (p->dpdk_v.copy_mode == DPDK_COPY_MODE_IPS &&
+                            !PacketTestAction(p, ACTION_DROP)))
+#if defined(RTE_LIBRTE_I40E_PMD) || defined(RTE_LIBRTE_IXGBE_PMD) || defined(RTE_LIBRTE_ICE_PMD)
+            && !(PKT_IS_ICMPV6(p) && p->icmpv6h->type == 143)
+#endif
+    ) {
+        retval =
+                rte_eth_tx_burst(p->dpdk_v.out_port_id, p->dpdk_v.out_queue_id, &p->dpdk_v.mbuf, 1);
+        if (unlikely(retval < 1)) {
+            // sometimes a repeated transmit can help to send out the packet
+            rte_delay_us(DPDK_BURST_TX_WAIT_US);
+            retval = rte_eth_tx_burst(
+                    p->dpdk_v.out_port_id, p->dpdk_v.out_queue_id, &p->dpdk_v.mbuf, 1);
+            if (unlikely(retval < 1)) {
+                SCLogDebug("Unable to transmit the packet on port %u queue %u",
+                        p->dpdk_v.out_port_id, p->dpdk_v.out_queue_id);
+                rte_pktmbuf_free(p->dpdk_v.mbuf);
+                p->dpdk_v.mbuf = NULL;
+            }
```

Comment:
I think you've meant if `retval > 0`. 
If the packet is sent successfully, then the function `rte_eth_tx_burst` will free the mbuf back to the pool. The return value of the function is the number of successfully sent packets. As I am only sending one packet (the last argument of the tx burst function), the returned value can be either 0 (in case of failure) or 1. If 1 is returned then the packet is sent and freed, if the return value is 0 then we need to free the packet

Reply:
I've added a comment there